pkg/crawler/crawler.go

```diff
@@ -0,0 +1,245 @@
+package crawler
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+	"github.com/aquasecurity/trivy-java-db/metadata"
+	"github.com/aquasecurity/trivy-java-db/pkg/db"
+	"github.com/aquasecurity/trivy-java-db/pkg/types"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/hashicorp/go-retryablehttp"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/xerrors"
+	"k8s.io/utils/clock"
+)
+
+const mavenRepoURL = "https://repo.maven.apache.org/maven2/"
+
+type Crawler struct {
+	wg             sync.WaitGroup
+	urlCh          chan string
+	indexCh        chan *types.Index
+	saveIndexesWG  sync.WaitGroup
+	tickerDuration time.Duration
+	limit          *semaphore.Weighted
+	client         *retryablehttp.Client
+	clock          clock.Clock
+}
+
+type Option struct {
+	Limit int64
+}
+
+func NewCrawler(opt Option) Crawler {
+	client := retryablehttp.NewClient()
+	client.Logger = nil
+	return Crawler{
+		urlCh:          make(chan string, opt.Limit*10),
+		indexCh:        make(chan *types.Index, opt.Limit),
+		tickerDuration: 500 * time.Millisecond,
+		limit:          semaphore.NewWeighted(opt.Limit),
+		client:         client,
+		clock:          clock.RealClock{},
+	}
+}
+
+func (c *Crawler) Crawl(ctx context.Context) error {
+	errCh := make(chan error)
+	defer close(errCh)
+
+	// Add a root url
+	c.urlCh <- mavenRepoURL
+	c.wg.Add(1)
+
+	go func() {
+		c.wg.Wait()
+		close(c.indexCh)
+		close(c.urlCh)
+		c.saveIndexesWG.Wait()
+	}()
+
+	go func() { // function for saving indexes in the database
+		c.saveIndexesWG.Add(1)
+		defer c.saveIndexesWG.Done()
+		ticker := time.NewTicker(c.tickerDuration)
+		defer ticker.Stop()
+		var indexes []*types.Index
+		for {
+			select {
+			case <-ticker.C: // save all indexes taken every c.tickerDuration
+				if len(indexes) > 0 {
+					db.InsertIndexes(indexes)
+					indexes = []*types.Index{} // clear array after saving to db
+				}
+			case index, ok := <-c.indexCh: // get index from chanel and save to array
+				if !ok {
+					return
+				}
+				indexes = append(indexes, index)
+			}
+		}
+	}()
+
+	var count int
+loop:
+	for {
+		select {
+		case url, ok := <-c.urlCh:
+			count++
+			if count%1000 == 0 {
+				log.Printf("Count: %d", count)
+			}
+			if !ok { // channel is closed
+				// save metadata
```

Comment:
We should put it outside of the for-loop.

Reply:
Done